trie/zk/merkle_tree.go

```diff
@@ -0,0 +1,479 @@
+	// It exists to decode the HashNode into the original TreeNode, so it is not needed if the HashNode cannot be created (e.g, rootNode is an EmptyNode).
+	findBlobByHash func(hash []byte) ([]byte, error)
+}
+
+func NewMerkleTreeFromHash(hash *zkt.Hash, findBlobByHash func(hash []byte) ([]byte, error)) (*MerkleTree, error) {
+	if bytes.Equal(hash.Bytes(), zkt.HashZero.Bytes()) {
+		return NewEmptyMerkleTree(), nil
+	}
+	blob, err := findBlobByHash(hash[:])
+	if err != nil {
+		return nil, err
+	}
+	rootNode, err := NewTreeNodeFromBlob(blob)
+	if err != nil {
+		return nil, err
+	}
+	setNodeHash(rootNode, hash)
+	return NewMerkleTreeFromNode(rootNode).WithNodeBlobFinder(findBlobByHash), nil
+}
+
+func NewMerkleTreeFromNode(node TreeNode) *MerkleTree { return NewMerkleTree(node) }
+func NewEmptyMerkleTree() *MerkleTree                 { return NewMerkleTree(EmptyNodeValue) }
+func NewMerkleTree(node TreeNode) *MerkleTree {
+	return &MerkleTree{rootNode: node, maxLevels: trie.NodeKeyValidBytes * 8}
+}
+
+func (t *MerkleTree) RootNode() TreeNode                      { return t.rootNode }
+func (t *MerkleTree) MaxLevels() int                          { return t.maxLevels }
+func (t *MerkleTree) WithMaxLevels(maxLevels int) *MerkleTree { t.maxLevels = maxLevels; return t }
+
+func (t *MerkleTree) WithNodeBlobFinder(findBlobByHash func(hash []byte) ([]byte, error)) *MerkleTree {
+	t.findBlobByHash = findBlobByHash
+	return t
+}
+
+// WithZkTrieCompatibilityMode hash collisions are checked every time the tree is updated.
+//
+// Whenever [trie.ZkTrie] is updated, it computes the hash collisions of all nodes on the path to the leaf node.
+// On the other hand, MerkleTree only checks for hash collisions when the ComputeAllNodeHash function is called.
+//
+// Because of this implementation difference, the root hash will be different if a hash collision occurs during a tree update.
+//
+// Since the main logic uses traditional [trie.ZkTrie], we recommend using unsafe mode only in specific cases.
+func (t *MerkleTree) WithZkTrieCompatibilityMode() *MerkleTree { return t.enableHashCollisionCheck() }
+
+// Hash returns the root hash of the MerkleTree.
+// It does not write to the database and can be used even if the MerkleTree doesn't have one.
+func (t *MerkleTree) Hash() []byte {
+	if err := t.ComputeAllNodeHash(nil); err != nil {
+		log.Error("Failed to compute hash", "err", err)
+		return []byte{}
+	}
+	return t.rootNode.Hash().Bytes()
+}
+
+// ComputeAllNodeHash Compute the node hash, and call the handleDirtyNode function.
+// handleDirtyNode is only called if the node hash is newly computed.
+func (t *MerkleTree) ComputeAllNodeHash(handleDirtyNode func(dirtyNode TreeNode) error) error {
+	if handleDirtyNode != nil && len(t.dirtyNode) != 0 {
+		for _, node := range t.dirtyNode {
+			if err := handleDirtyNode(node); err != nil {
+				return err
+			}
+		}
+	}
+	return t.enableHashCollisionCheck().checkNodeHashConflict(t.rootNode, false, handleDirtyNode)
```

Comment:
It seems that you assume that `enableHashCollisionCheck` is on all the times. because you don't check `t.dirtyNode == nil` at the beginning of the function and that means you can just call ` t.checkNodeHashConflict(t.rootNode, false, handleDirtyNode)`